sdks/java/extensions/sql/zetasql/src/main/java/org/apache/beam/sdk/extensions/sql/zetasql/translation/ExpressionConverter.java

```diff
@@ -786,15 +786,39 @@ private RexNode convertSimpleValueToRexNode(TypeKind kind, Value value) {
                     ZetaSqlCalciteTranslationUtils.toSimpleRelDataType(kind, rexBuilder()));
         break;
       case TYPE_DOUBLE:
+        // Cannot simply call makeApproxLiteral() for ZetaSQL DOUBLE type because positive infinity,
+        // negative infinity and Nan cannot be directly converted to BigDecimal. So we create three
+        // wrapper functions here for these three cases such that we can later recognize it and
+        // customize its unparsing in BeamBigQuerySqlDialect.
         double val = value.getDoubleValue();
-        if (Double.isInfinite(val) || Double.isNaN(val)) {
-          throw new UnsupportedOperationException("Does not support Infinite or NaN literals.");
+        if (Double.isInfinite(val) && val > 0) {
+          ret =
+              rexBuilder()
+                  .makeCall(
+                      SqlOperators.createOtherKindSqlFunction(
+                          BeamBigQuerySqlDialect.DOUBLE_POSITIVE_INF_FUNCTION,
+                          ZetaSqlCalciteTranslationUtils.toCalciteTypeName(kind)));
+        } else if (Double.isInfinite(val) && val < 0) {
+          ret =
+              rexBuilder()
+                  .makeCall(
+                      SqlOperators.createOtherKindSqlFunction(
+                          BeamBigQuerySqlDialect.DOUBLE_NEGATIVE_INF_FUNCTION,
+                          ZetaSqlCalciteTranslationUtils.toCalciteTypeName(kind)));
+        } else if (Double.isNaN(val)) {
+          ret =
+              rexBuilder()
+                  .makeCall(
+                      SqlOperators.createOtherKindSqlFunction(
+                          BeamBigQuerySqlDialect.DOUBLE_NAN_FUNCTION,
+                          ZetaSqlCalciteTranslationUtils.toCalciteTypeName(kind)));
+        } else {
+          ret =
+              rexBuilder()
+                  .makeApproxLiteral(
+                      new BigDecimal(val),
+                      ZetaSqlCalciteTranslationUtils.toSimpleRelDataType(kind, rexBuilder()));
```

Comment:
I could create 2 local parameters: `String wrapperFun` (null for the last case) and `RelDataType returnType` to simplify this big if-else block.

Reply:
I am not quiet sure how to merge the first three blocks with the last one, because the first three use `resBuilder().makeCall()` while the last one uses `rexBuilder().makeApproxLiteral()`. Also, `returnType` seems can only be used in the last block.

Reply:
I meant
```
String wrapperFun = null;
if (val == Double.POSITIVE_INFINITY) {
  wrapperFun = BeamBigQuerySqlDialect.DOUBLE_POSITIVE_INF_FUNCTION;
} else if (val == Double.NEGATIVE_INFINITY) {
  BeamBigQuerySqlDialect.DOUBLE_NEGATIVE_INF_FUNCTION
} else if (Double.isNan(val)) {
  wrapperFun = BeamBigQuerySqlDialect.DOUBLE_NAN_FUNCTION;
}
RelDataType returnType = ZetaSqlCalciteTranslationUtils.toSimpleRelDataType(kind, rexBuilder());
if (wrapperFun == null) {
  ret = rexBuilder().makeApproxLiteral(new BigDecimal(val), returnType);
} else {
  ret = rexBuilder().makeCall(SqlOperators.createOtherKindSqlFunction(wrapperFun, returnType.getSqlTypeName()));
}
```

Reply:
That's seems good idea. Got it.

Reply:
Done.